File: Client_Start/Assets/LuaFramework/Lua/MyLua/Libs/Table/TableSys.lua
--[[
 * @brief 添加一个表的步骤总共分 4 步
 * // 添加一个表的步骤一
 * // 添加一个表的步骤二
 * // 添加一个表的步骤三
 * // 添加一个表的步骤四
]]

MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "TableSys";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.m_dicTable = GlobalNS.new(GlobalNS.MDictionary);
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_OBJECT, GlobalNS.new(GlobalNS.TableBase, "ObjectBase_client.bytes", "ObjectBase_client"));
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_CARD, GlobalNS.new(GlobalNS.TableBase, "CardBase_client.bytes", "CardBase_client"));
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_SKILL, GlobalNS.new(GlobalNS.TableBase, "SkillBase_client.bytes", "SkillBase_client"));    -- 添加一个表的步骤三
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_JOB, GlobalNS.new(GlobalNS.TableBase, "proBase_client.bytes", "proBase_client"));
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_SPRITEANI, GlobalNS.new(GlobalNS.TableBase, "FrameAni_client.bytes", "FrameAni_client"));
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_RACE, GlobalNS.new(GlobalNS.TableBase, "RaceBase_client.bytes", "RaceBase_client"));
    self.m_dicTable:Add(GlobalNS.TableID.TABLE_STATE, GlobalNS.new(GlobalNS.TableBase, "StateBase_client.bytes", "StateBase_client"));
end

-- 返回一个表
function getTable(tableID)
	local table = self.m_dicTable:value(tableID);
	if (nil == table) then
		self:loadOneTable(tableID);
		table = self.m_dicTable:value(tableID);
	end
	return table.m_List;
end

-- 返回一个表中一项，返回的时候表中数据全部加载到 Item 中
function M:getItem(tableID, itemID)
    local table = self.m_dicTable:value(tableID);
    if (nil == table.m_byteBuffer) then
		self:loadOneTable(tableID);
		table = self.m_dicTable:value(tableID);
	end
    local ret = self:findDataItem(table, itemID);

    if (nil ~= ret and nil == ret.m_itemBody) then
        self:loadOneTableOneItemAll(tableID, table, ret);
    end

    if (nil == ret) then
        -- 日志
    end

	return ret;
end

-- 加载一个表
function M:loadOneTable(tableID)
	local table = self.m_dicTable:value(tableID);
	
	local auxBytesLoader = GlobalNS.new(GlobalNS.AuxBytesLoader);
	local path = GlobalNS.UtilPath.CombineTwo(GCtx.m_config.m_pathLst[GlobalNS.ResPathType.ePathTablePath], table.m_resName);
	
	auxBytesLoader:syncLoad(path, self, self.onLoadEventHandle);
end

-- 加载一个表完成
function M:onLoadEventHandle(dispObj)
    self.m_res = dispObj;
    if (self.m_res:hasSuccessLoaded()) then
        GCtx.mLogSys:log(self.m_res:getLogicPath(), GlobalNS.LogTypeId.eLogCommon);

        local bytes = self.m_res:getBytes();
        if (nil ~= bytes) then
            self.m_byteArray = GlobalNS.CSSystem.buildByteBuffer();
            self.m_byteArray:clear();
            self.m_byteArray:writeBytes(bytes, 0, bytes.Length, true);
            self.m_byteArray:setPos(0);
            self:readTable(self:getTableIDByPath(self.m_res:getLogicPath()), self.m_byteArray);
        end
    elseif (self.m_res:hasFailed()) then
		GCtx.mLogSys:log(self.m_res:getLogicPath(), GlobalNS.LogTypeId.eLogCommon);
    end

    -- 卸载资源
    self.m_res:dispose();
	self.m_res = nil;
end

-- 根据路径查找表的 ID
function M:getTableIDByPath(path)
	local tablePath = "";
	
    for key, value in pairs(self.m_dicTable:getData()) do
		tablePath = GlobalNS.UtilPath.CombineTwo(GCtx.m_config.m_pathLst[GlobalNS.ResPathType.ePathTablePath], value.m_resName);
        if (tablePath == path) then
            return key;
        end
    end

    return 0;
end

-- 加载一个表中一项的所有内容
function M:loadOneTableOneItemAll(tableID, table, itemBase)
    if (GlobalNS.TableID.TABLE_OBJECT == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableObjectItemBody);
    elseif (GlobalNS.TableID.TABLE_CARD == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableCardItemBody);
    elseif (GlobalNS.TableID.TABLE_SKILL == tableID) then  -- 添加一个表的步骤四
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableSkillItemBody);
    elseif (GlobalNS.TableID.TABLE_JOB == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableJobItemBody);
    elseif (GlobalNS.TableID.TABLE_SPRITEANI == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableSpriteAniItemBody);
    elseif (GlobalNS.TableID.TABLE_RACE == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableRaceItemBody);
    elseif (GlobalNS.TableID.TABLE_STATE == tableID) then
        itemBase.parseBodyByteBuffer(table.m_byteBuffer, itemBase.m_itemHeader.m_offset, GlobalNS.TableStateItemBody);
    end
end

-- 获取一个表的名字
function M:getTableName(tableID)
	local table = self.m_dicTable:value(tableID);
	if (nil ~= table) then
		return table.m_tableName;
	end
	return "";
end

-- 读取一个表，仅仅读取表头
function M:readTable(tableID, bytes)
    local table = self.m_dicTable:value(tableID);
    table.m_byteBuffer = bytes;

    --bytes:setEndian(GlobalNS.EEndian.eLITTLE_ENDIAN);
	bytes:setEndian(GlobalNS.CSSystem.EEndian.eLITTLE_ENDIAN);
    local count = 0;
    bytes:readUnsignedInt32(count);
    local i = 0;
    local item = nil;
    while(i < count) do
        item = GlobalNS.new(GlobalNS.TableItemBase);
        item:parseHeaderByteBuffer(bytes);
        table.m_List:Add(item);
		
		i = i + 1
    end
end

-- 查找表中的一项
function M:findDataItem(table, id)
	local size = table.m_List:Count();
	local low = 0;
	local high = size - 1;
	local middle = 0;
	local idCur = 0;
	
	while (low <= high) do
		middle = (low + high) / 2;
        idCur = table.m_List:at(middle).m_itemHeader.m_uID;
		if (idCur == id) then
			break;
		end
		if (id < idCur)then
			high = middle - 1;
		else
			low = middle + 1;
		end
	end
	
	if (low <= high) then
        return table.m_List:at(middle);
	end
	
	return nil;
end

return M;

File: PlaneWar/Client_1/Assets/Scripts/Libs/Table/TableSys.cs
using System.Collections.Generic;
using System.IO;

namespace SDK.Lib
{
    /**
     * @brief 添加一个表的步骤总共分 4 步
     * // 添加一个表的步骤一
     * // 添加一个表的步骤二
     * // 添加一个表的步骤三
     * // 添加一个表的步骤四
     */
    public class TableSys
	{
        private MDictionary<TableID, TableBase> mDicTable;
		private ResItem mRes;
        private ByteBuffer mByteArray;

		public TableSys()
		{
			mDicTable = new MDictionary<TableID, TableBase>();
            mDicTable[TableID.TABLE_OBJECT] = new TableBase("ObjectBase_client.bytes", "ObjectBase_client");
            mDicTable[TableID.TABLE_CARD] = new TableBase("CardBase_client.bytes", "CardBase_client");
            mDicTable[TableID.TABLE_SKILL] = new TableBase("SkillBase_client.bytes", "SkillBase_client");    // 添加一个表的步骤三
            mDicTable[TableID.TABLE_JOB] = new TableBase("proBase_client.bytes", "proBase_client");
            mDicTable[TableID.TABLE_SPRITEANI] = new TableBase("FrameAni_client.bytes", "FrameAni_client");
            mDicTable[TableID.TABLE_RACE] = new TableBase("RaceBase_client.bytes", "RaceBase_client");
            mDicTable[TableID.TABLE_STATE] = new TableBase("StateBase_client.bytes", "StateBase_client");
		}

        // 返回一个表
        public List<TableItemBase> getTable(TableID tableID)
		{
			TableBase table = mDicTable[tableID];
			if (null == table)
			{
				loadOneTable(tableID);
				table = mDicTable[tableID];
			}
			return table.mList;
		}
		
        // 返回一个表中一项，返回的时候表中数据全部加载到 Item 中
        public TableItemBase getItem(TableID tableID, uint itemID)
		{
            TableBase table = mDicTable[tableID];
            if (null == table.mByteBuffer)
			{
				loadOneTable(tableID);
				table = mDicTable[tableID];
			}
            TableItemBase ret = TableSys.findDataItem(table, itemID);

            if (null != ret && null == ret.mItemBody)
            {
                loadOneTableOneItemAll(tableID, table, ret);
            }

            if (null == ret)
            {
                if (MacroDef.ENABLE_LOG)
                {
                    Ctx.mInstance.mLogSys.log(string.Format("table name: {0}, table Item {1} load failed", (int)tableID, itemID));
                }
            }

			return ret;
		}
		
        // 加载一个表
		public void loadOneTable(TableID tableID)
		{
			TableBase table = mDicTable[tableID];

            LoadParam param = Ctx.mInstance.mPoolSys.newObject<LoadParam>();
            param.setPath(Path.Combine(Ctx.mInstance.mCfg.mPathLst[(int)ResPathType.ePathTablePath], table.mResName));
            param.mLoadEventHandle = onLoadEventHandle;
            param.mLoadNeedCoroutine = false;
            param.mResNeedCoroutine = false;
            Ctx.mInstance.mResLoadMgr.loadAsset(param);
            Ctx.mInstance.mPoolSys.deleteObj(param);
		}

        // 加载一个表完成
        public void onLoadEventHandle(IDispatchObject dispObj)
        {
            mRes = dispObj as ResItem;
            if (mRes.refCountResLoadResultNotify.resLoadState.hasSuccessLoaded())
            {
                Ctx.mInstance.mLogSys.debugLog_1(LangItemID.eItem0, mRes.getLoadPath());

                byte[] bytes = mRes.getBytes("");
                if (bytes != null)
                {
                    mByteArray = Ctx.mInstance.mFactoryBuild.buildByteBuffer();
                    mByteArray.clear();
                    mByteArray.writeBytes(bytes, 0, (uint)bytes.Length);
                    mByteArray.setPos(0);
                    readTable(getTableIDByPath(mRes.getLogicPath()), mByteArray);
                }
            }
            else if (mRes.refCountResLoadResultNotify.resLoadState.hasFailed())
            {
                Ctx.mInstance.mLogSys.debugLog_1(LangItemID.eItem1, mRes.getLoadPath());
            }

            // 卸载资源
            Ctx.mInstance.mResLoadMgr.unload(mRes.getResUniqueId(), onLoadEventHandle);
        }

        // 根据路径查找表的 ID
        protected TableID getTableIDByPath(string path)
        {
            foreach (KeyValuePair<TableID, TableBase> kv in mDicTable.getData())
            {
                if (Ctx.mInstance.mCfg.mPathLst[(int)ResPathType.ePathTablePath] + kv.Value.mResName == path)
                {
                    return kv.Key;
                }
            }

            return 0;
        }

        // 加载一个表中一项的所有内容
        public void loadOneTableOneItemAll(TableID tableID, TableBase table, TableItemBase itemBase)
        {
            if (TableID.TABLE_OBJECT == tableID)
            {
                itemBase.parseBodyByteBuffer<TableObjectItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_CARD == tableID)
            {
                itemBase.parseBodyByteBuffer<TableCardItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_SKILL == tableID)  // 添加一个表的步骤四
            {
                itemBase.parseBodyByteBuffer<TableSkillItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_JOB == tableID)
            {
                itemBase.parseBodyByteBuffer<TableJobItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_SPRITEANI == tableID)
            {
                itemBase.parseBodyByteBuffer<TableSpriteAniItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_RACE == tableID)
            {
                itemBase.parseBodyByteBuffer<TableRaceItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
            else if (TableID.TABLE_STATE == tableID)
            {
                itemBase.parseBodyByteBuffer<TableStateItemBody>(table.mByteBuffer, itemBase.mItemHeader.mOffset);
            }
        }
		
        // 获取一个表的名字
		public string getTableName(TableID tableID)
		{
			TableBase table = mDicTable[tableID];
			if (null != table)
			{
				return table.mTableName;
			}			
			return "";
		}

        // 读取一个表，仅仅读取表头
        private void readTable(TableID tableID, ByteBuffer bytes)
        {
            TableBase table = mDicTable[tableID];
            table.mByteBuffer = bytes;

            bytes.setEndian(EEndian.eLITTLE_ENDIAN);
            uint len = 0;
            bytes.readUnsignedInt32(ref len);
            uint i = 0;
            TableItemBase item = null;
            for (i = 0; i < len; i++)
            {
                //if (TableID.TABLE_OBJECT == tableID)
                //{
                //    item = new TableItemObject();
                //}
                item = new TableItemBase();
                item.parseHeaderByteBuffer(bytes);
                // 加载完整数据
                //loadOneTableOneItemAll(tableID, table, item);
                //if (TableID.TABLE_OBJECT == tableID)
                //{
                    //item.parseAllByteBuffer<TableObjectItemBody>(bytes);
                //}
                table.mList.Add(item);
            }
        }

        // 查找表中的一项
        static public TableItemBase findDataItem(TableBase table, uint id)
		{
			int size = table.mList.Count;
			int low = 0;
			int high = size - 1;
			int middle = 0;
			uint idCur = 0;
			
			while (low <= high)
			{
				middle = (low + high) / 2;
                idCur = table.mList[middle].mItemHeader.mId;
				if (idCur == id)
				{
					break;
				}
				if (id < idCur)
				{
					high = middle - 1;
				}
				else
				{
					low = middle + 1;
				}
			}
			
			if (low <= high)
			{
                return table.mList[middle];
			}
			return null;
		}
	}
}